kernel/timeout.c

```diff
@@ -96,7 +96,8 @@ int _abort_timeout(struct _timeout *to)
 	int ret = _INACTIVE;
 
 	LOCKED(&timeout_lock) {
-		if (to->dticks != _INACTIVE) {
+		if (sys_dnode_is_linked(&to->node)
+		    && (to->dticks != _INACTIVE)) {
```

Comment:
@andyross Please review.  In particular, I was surprised that it's possible for a timeout structure to be linked while it's marked inactive.  That could happen legitimately if the timeout is linked into a different list (e.g. of unused timeouts), but I'm not seeing how that could be.

If the check for `dticks` is removed then several sanity checks fail, including `samples/cpp_synchronization/test`.

Reply:
[pabigot/zephyr:pr/20181230b](https://github.com/pabigot/zephyr/tree/pr/20181230b) is this branch rebased on #12204 after removing your workaround.  With the following change it fails `cpp_synchronization`.  So the problem is independent of whether #12204 is merged.
```
diff --git a/kernel/timeout.c b/kernel/timeout.c
index ffe6e78039..6ea334f461 100644
--- a/kernel/timeout.c
+++ b/kernel/timeout.c
@@ -96,8 +96,8 @@ int _abort_timeout(struct _timeout *to)
        int ret = _INACTIVE;
 
        LOCKED(&timeout_lock) {
-               if (sys_dnode_is_linked(&to->node)
-                   && (to->dticks != _INACTIVE)) {
+               if (sys_dnode_is_linked(&to->node)) {
+                       printk("remove %p, %d\n", &to->node, to->dticks);
                        remove_timeout(to);
                        ret = 0;
                }
```

Reply:
I found a "hidden" `_init_timeout` function that needed to be updated so a stack-allocated `k_timer` in the C++ sample didn't start with bogus pointers.  So the secondary check is no longer present, and it'd probably be possible to eliminate `_INACTIVE` as a flag value to detect disabled timeout structures.